_posts/2025-04-16-tha-various-tiers-of-apache-kafka-tiered-storage.md

```diff
@@ -0,0 +1,462 @@
+  # ...
+```
+
+And in all `KafkaNodePool` resources with the `broker` role, we have to mount the NFS volume in `.spec.template` using the additional volumes feature.
+The `KafkaNodePool` resources that have only the `controller` role do not need the volume as they won't use tiered storage.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: broker
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  roles:
+    - broker
+  # ...
+  template:
+    pod:
+      volumes:
+        - name: tiered-storage
+          persistentVolumeClaim:
+            claimName: tiered-storage-nfs
+    kafkaContainer:
+      volumeMounts:
+        - name: tiered-storage
+          mountPath: /mnt/tiered-storage/
+```
+
+Once the Kafka cluster is ready, we can create a Kafka topic named `tiered-storage-test`.
+Keep in mind that in order to use the tiered storage, you have to enable in the topic as well:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaTopic
+metadata:
+  name: tiered-storage-test
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 10
+  replicas: 3
+  config:
+    min.insync.replicas: 2
+    retention.bytes: 107374182400 # ~100 Gi
+    retention.ms: 604800000 # 7 days
+    segment.bytes: 10485760 # ~10MB
+    file.delete.delay.ms: 1000
+    # Tiered storage configuration
+    remote.storage.enable: true
+    local.retention.ms: 60000 # 1 minute
+    local.retention.bytes: 50000000 # 50 MB
+```
+
+To make it easier to see how the tiered storage works, the YAML above tunes some of the configuration options:
+* Uses small segment size
+* Keeps the local retention very short so that we can better see the log segments being offloaded to the remote storage tier
+
+In real-world deployments, these values are typically higher.
+But the actual values might differ based on your use-case.
+So make sure to tune them accordingly.
+
+Finally, with the topic ready and with tiered storage enabled, we can start producing some messages.
+To demonstrate the tiered storage functionality, we can use a Kubernetes Job to produce a large amount of messages to our topic:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  labels:
+    app: kafka-producer
+  name: kafka-producer
+spec:
+  parallelism: 5
+  completions: 5
+  backoffLimit: 1
+  template:
+    metadata:
+      name: kafka-producer
+      labels:
+        app: kafka-producer
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: kafka-producer
+        image: quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
+        command: [ "bin/kafka-producer-perf-test.sh" ]
+        args: [ "--topic", "tiered-storage-test", "--throughput", "1000000000", "--num-records", "1000000000", "--producer-props", "acks=all", "bootstrap.servers=my-cluster-kafka-bootstrap:9092", "--record-size", "1000" ]
+```
+
+The Job above starts 5 parallel producers, each sending `1000000000` records with 1000 bytes each to our `tiered-storage-test` topic.
```

Comment:
Isn't this 4.5 TiB, which is much larger than the 100 GiB of the test topic retention? Do we need to send that much data for this example?

Reply:
Its just to make sure there is some traffic going.

Reply:
Ok, but if the retention kicks in, you will get different results than the one described at the end of the post and this may leave some readers confused.